vault/resource_database_secret_backend_connection.go

```diff
@@ -1786,9 +1806,29 @@ func setDatabaseConnectionDataWithUserPass(d *schema.ResourceData, prefix string
 	// the old password in the update request would break the connection config. Thus we only send it,
 	// if it actually changed to still support updating it for non-rotated cases.
 	passwordKey := prefix + consts.FieldPassword
-	if v, ok := d.GetOk(passwordKey); ok {
-		if d.IsNewResource() || d.HasChange(passwordKey) {
+	passwordWriteOnlyVersionKey := prefix + consts.FieldPasswordWOVersion
+	if d.IsNewResource() || d.HasChange(passwordKey) || d.HasChange(passwordWriteOnlyVersionKey) {
+		if v, ok := d.GetOk(passwordKey); ok && v != nil {
+			log.Printf("[DEBUG] using persisted password; please use new write-only attributes for security")
 			data[consts.FieldPassword] = v.(string)
+		} else {
+			engineName, engineIdx, err := databaseEngineNameAndIndexFromPrefix(prefix)
+			if err != nil {
+				// this should not happen, since we control how the prefix is created
+				log.Fatalf("[ERROR] invalid prefix %q for database connection: %s", prefix, err)
```

Comment:
If this is fatal, should we panic?

Reply:
Panic is a better UX for the user to determine exactly what went wrong. Switched to be a panic here instead of `os.Exit`